Refactor subbridge pallets

tolak opened the pull request:
# ⚠️Breaking Change⚠️

This PR contains old pallet rename and removing, we must make sure downstream(like Chanbridge relayer) and migration get ready before merge it.

## Breaking changes:

- Move `pallet_assets_wrapper` outside `xtransfer`, and rename to `assets_registry`
  - Remove storage `IdByAssets`, have `IdByLocation` instead
  - Remove storage `AssetByIds` and `AssetByResourceIds`, have `IdByResourceId` instead
- Combine `pallet_bridge` and `pallet_bridge_transfer` into `chainbridge`
  - Move storage `BridgeFee` into new pallet `chainbridge`
  - Move all storage in `pallet_bridge` into new pallet `chainbridge`
- Change `BridgeTransfer.transfer` to `ChainBridge.handle_fungible_transfer`(**Chainbridge relayer must follow**)
- `fee` will be deducted from withdrawn `amount` when transfer to solo chain through Chainbridge

## Important changes:

- Combine all crosschain transfer extrinsic call to `XTransfer.transfer()`, incluing transfer native and asset through XCM or bridges.
- Support multi-bridge configuration in pallet xtransfer, e.g. if we have `xtransfer::config::Bridge = (BridgeA, BridgeB, BridgeC)`, it will try transfer by the sequence of `BridgeA`, `BridgeB`, and `BridgeC`, untill one of them get success. If all failed, error will returned.
- Use `MultiAsset` as the only way to describe an asset, use `MultiLocation` as the only way to describe a destination.
- All pallets under xtransfer module, e.g. `chainbridge`, `xcm_transfer`, use same mock runtime(which is based on xcm-simulator)
- Introduce `BridgeTransact` trait, every bridge pallet has its own `BridgeTransactImpl` struct to impl this trait, which would be configured to xtransfer

h4x3rotab commented on pallets/xtransfer/src/chainbridge.rs:
```diff
@@ -0,0 +1,2408 @@
+
+		fn can_send_data(_data: &Vec<u8>, _dest: MultiLocation) -> bool {
+			// TODO: impl
+			return false;
```
```suggestion
			false
```

tolak replied:
Done

h4x3rotab reviewed (changes requested):
General questions:

1. How about changing `xtransfer` to `subbridge`?
2. How about calling the transfer bridge implementation something like "xcm_bridge" or "xcm_bridge_plugin" to imply it's one pluggable bridge in SubBridge?
3. What will happen if the bridge routing fails? Is it possible for the original sender chain to claim the stuck assets back?
4. Where is the received crosschain assets handled?

tolak commented on pallets/xtransfer/src/chainbridge.rs:
```diff
@@ -0,0 +1,2408 @@
+				(Some((asset_location, _)), Some((dest_id, _))) => {
+					let rid = asset_location.clone().into_rid(dest_id);
+					// Verify if dest chain has been whitelisted
+					if Self::chain_whitelisted(dest_id) == false {
```
Done

tolak commented on pallets/xtransfer/src/chainbridge.rs:
```diff
@@ -0,0 +1,2408 @@
+		}
+
+		fn extract_dest(dest: &MultiLocation) -> Option<(BridgeChainId, Vec<u8>)> {
+			return match (dest.parents, &dest.interior) {
```
Done

tolak commented on pallets/xtransfer/src/chainbridge.rs:
```diff
@@ -0,0 +1,2408 @@
+		}
+
+		fn extract_fungible(asset: MultiAsset) -> Option<(MultiLocation, u128)> {
+			return match (asset.fun, asset.id) {
+				(Fungible(amount), Concrete(location)) => Some((location, amount)),
+				_ => None,
+			};
```
Done

tolak commented on pallets/xtransfer/src/chainbridge.rs:
```diff
@@ -0,0 +1,2408 @@
+		}
+
+		pub fn estimate_fee_in_pha(dest_id: BridgeChainId, amount: u128) -> Option<u128> {
+			return match Self::bridge_fee(dest_id) {
+				Some((min_fee, fee_scale)) => {
+					let fee_estimated = amount
+						.saturating_mul(fee_scale as u128)
+						.saturating_div(1000u128);
+					if fee_estimated > min_fee {
+						Some(fee_estimated)
+					} else {
+						Some(min_fee)
+					}
+				}
+				_ => None,
+			};
```
Done

tolak commented on pallets/xtransfer/src/chainbridge.rs:
```diff
@@ -0,0 +1,2408 @@
+
+	#[pallet::storage]
+	#[pallet::getter(fn bridge_fee)]
+	pub type BridgeFee<T: Config> = StorageMap<_, Twox64Concat, BridgeChainId, (u128, u32)>;
```
Yes, will do it

tolak commented on pallets/assets-registry/src/lib.rs:
```diff
@@ -445,62 +469,212 @@ pub mod pallet {
+				};
+
+				// Should be failed if origin is from sudo user
+				assert_err!(
```
Done

tolak commented on pallets/assets-registry/src/lib.rs:
```diff
@@ -445,62 +469,212 @@ pub mod pallet {
+					interior: X1(Parachain(1)),
+				};
+
+				// Should be failed if origin is from sudo user
```
Done

tolak commented on pallets/assets-registry/src/lib.rs:
```diff
@@ -203,12 +194,12 @@ pub mod pallet {
+			location: MultiLocation,
 		},
 		/// Asset is unregisterd.
 		AssetUnRegistered {
```
Done

tolak commented on pallets/xtransfer/src/traits.rs:
```diff
@@ -0,0 +1,122 @@
+            }
+        )* );
+		let last_error = last_error.unwrap_or(DispatchError::Other("TransactFailed"));
+		log::trace!(target: "BridgeTransact::transfer_generic", "last_error: {:?}", last_error);
```
Done

tolak commented on pallets/xtransfer/src/traits.rs:
```diff
@@ -0,0 +1,122 @@
+            }
+        )* );
+		let last_error = last_error.unwrap_or(DispatchError::Other("TransactFailed"));
+		log::trace!(target: "BridgeTransact::transfer_nonfungible", "last_error: {:?}", last_error);
```
Done

tolak commented on pallets/xtransfer/src/traits.rs:
```diff
@@ -0,0 +1,122 @@
+            }
+        )* );
+		let last_error = last_error.unwrap_or(DispatchError::Other("TransactFailed"));
+		log::trace!(target: "BridgeTransact::transfer_fungible", "last_error: {:?}", last_error);
```
Done

tolak commented on pallets/xtransfer/src/xcm_transfer.rs:
```diff
@@ -0,0 +1,600 @@
+		fn test_transfer_native_to_parachain() {
+			TestNet::reset();
+
+			let para_a_location: MultiLocation = MultiLocation {
```
Done

tolak commented on pallets/xtransfer/src/xcm_transfer.rs:
```diff
@@ -0,0 +1,600 @@
+		fn test_transfer_to_reserve_parachain() {
+			TestNet::reset();
+
+			let para_a_location: MultiLocation = MultiLocation {
```
Done

tolak commented on pallets/xtransfer/src/xcm_transfer.rs:
```diff
@@ -0,0 +1,600 @@
+		) -> DispatchResult {
+			log::trace!(
+				target: LOG_TARGET,
+				" Xcm fungible transfer, sender: {:?}, asset: {:?}, dest: {:?}, max_weight: {:?}.",
```
Done

tolak commented on pallets/xtransfer/src/xcm_transfer.rs:
```diff
@@ -0,0 +1,600 @@
+			}
+
+			if let Some((asset_location, amount)) = Self::extract_fungible(asset.clone()) {
+				// Verify if asset was registered if is not native.
```
Done

tolak commented on pallets/xtransfer/src/xcm_transfer.rs:
```diff
@@ -0,0 +1,600 @@
+			}
+
+			if let Some((asset_location, amount)) = Self::extract_fungible(asset.clone()) {
+				// Verify if asset was registered if is not native.
```
Done

tolak commented on pallets/xtransfer/src/xcm_transfer.rs:
```diff
@@ -0,0 +1,600 @@
+		<T as frame_system::Config>::AccountId: From<[u8; 32]> + Into<[u8; 32]>,
+	{
+		fn can_deposit_asset(asset: MultiAsset, dest: MultiLocation) -> bool {
+			// Only support transfer to relaychain or parachain
```
Yes, but xcm_transfer pallet only handler crosschain, transfer to local is not allowed here.

tolak commented on pallets/xtransfer/src/xcm_transfer.rs:
```diff
@@ -0,0 +1,600 @@
+
+		fn can_send_data(data: &Vec<u8>, dest: MultiLocation) -> bool {
+			// TODO: impl
+			return false;
```
Done

tolak commented on pallets/xtransfer/src/xcm_transfer.rs:
```diff
@@ -0,0 +1,600 @@
+
+	impl<T: Config> Pallet<T> {
+		fn extract_fungible(asset: MultiAsset) -> Option<(MultiLocation, u128)> {
+			return match (asset.fun, asset.id) {
+				(Fungible(amount), Concrete(location)) => Some((location, amount)),
+				_ => None,
+			};
```
Done

tolak commented on pallets/xtransfer/src/xcm_transfer.rs:
```diff
@@ -0,0 +1,600 @@
+			})
+		}
+
+		fn invert_based_reserve(
```
Done

tolak commented on pallets/xtransfer/src/xcm_transfer.rs:
```diff
@@ -0,0 +1,600 @@
+	}
+
+	impl<T: Config> XCMSession<T> {
+		fn kind(&self) -> Option<TransferType> {
+			let mut transfer_type = None;
+			ConcrateAsset::origin(&self.asset).map(|asset_reserve_location| {
+				if T::NativeAssetChecker::is_native_asset_location(&asset_reserve_location) {
+					transfer_type = Some(TransferType::FromNative);
+				} else if asset_reserve_location == self.dest_location {
+					transfer_type = Some(TransferType::ToReserve);
+				} else {
+					transfer_type = Some(TransferType::ToNonReserve);
+				}
+			});
+			transfer_type
+		}
```
Done

tolak commented on pallets/assets-registry/src/lib.rs:
```diff
@@ -163,35 +158,31 @@ pub mod pallet {
-	pub type IdByAssets<T: Config> =
-		StorageMap<_, Twox64Concat, XTransferAsset, <T as pallet_assets::Config>::AssetId>;
+	#[pallet::getter(fn location_to_id)]
+	pub type IdByLocation<T: Config> =
```
Done

tolak commented on pallets/assets-registry/src/lib.rs:
```diff
@@ -240,41 +231,70 @@ pub mod pallet {
+					&fund_account,
+					&recipient,
+					amount.into(),
+					ExistenceRequirement::KeepAlive,
```
Done

tolak commented on runtime/khala/src/lib.rs:
```diff
@@ -249,18 +251,18 @@ construct_runtime! {
-        XcmTransfer: pallet_xcm_transfer::{Pallet, Call, Event<T>, Storage} = 82,
+        // Bridges
+        ChainBridge: chainbridge::{Pallet, Call, Storage, Event<T>} = 80,
+        XcmTransfer: xcm_transfer::{Pallet, Event<T>, Storage} = 81,
+        XTransfer: xtransfer::{Pallet, Call, Storage, Event<T>} = 82,
```
Yes, will do this before upgrade

tolak commented on runtime/rhala/src/lib.rs:
```diff
@@ -813,6 +819,8 @@ pub type XcmOriginToTransactDispatchOrigin = (
     pub UnitWeightCost: Weight = 200_000_000;
     pub const MaxInstructions: u32 = 100;
+    pub ParaTreasuryAccount: AccountId = PalletId(*b"py/trsry").into_account();
+	pub ParaCheckingAccount: AccountId = PalletId(*b"py/check").into_account();
```
Done

tolak commented on runtime/rhala/src/lib.rs:
```diff
@@ -824,19 +832,187 @@ pub type Barrier = (
+    pub ExecutionPrices: Vec<(AssetId, u128)> = [
+        ExecutionPriceInKSM::get(),
+        ExecutionPriceInPHA::get(),
+        ExecutionPriceInLocalPHA::get(),
```
They both represent PHA, just to be compatible with XCM location converter

tolak commented on runtime/thala/src/lib.rs:
```diff
@@ -1432,31 +1426,32 @@ parameter_types! {
-    type TreasuryAccount = KhalaTreasuryAccount;
+    type ProposalLifetime = ProposalLifetime;
+    type NativeAssetChecker = helper::NativeAssetFilter<ParachainInfo>;
+	type NativeExecutionPrice = NativeExecutionPrice;
```
Done

tolak commented on pallets/xtransfer/src/xcm_transfer.rs:
```diff
@@ -0,0 +1,600 @@
+				// ParaA send it's own native asset to paraB
+				assert_ok!(bridge_impl.transfer_fungible(
+					ALICE.into(),
+					(Concrete(MultiLocation::new(0, Here)), Fungible(10u128)).into(),
```
We already have `NativeAssetChecker::native_asset_location()` and i suggest let XCM convert sibling parachain account from MultiLocation, we use soverign account only in test.

h4x3rotab replied:
OK. I meant to simplify the test code. It's fine to keep the explicit definition as well.

tolak commented on pallets/xtransfer/src/xtransfer.rs:
```diff
@@ -0,0 +1,619 @@
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	#[pallet::without_storage_info]
```
Done

tolak commented on pallets/xtransfer/src/fungible_adapter.rs:
```diff
@@ -0,0 +1,118 @@
+		NativeChecker: NativeAssetChecker,
+	> TransactAsset for XTransferAdapter<NativeAdapter, AssetsAdapter, Transactor, NativeChecker>
+{
+	fn can_check_in(_origin: &MultiLocation, what: &MultiAsset) -> XcmResult {
```
Yes since we haven't support it so far, it's better to reject them. Already remove those methods in fungible adapter, the default impl of Transact trait will throw an error.

tolak commented on pallets/xtransfer/src/fungible_adapter.rs:
```diff
@@ -0,0 +1,118 @@
+				} else {
+					AssetsAdapter::deposit_asset(what, who)?;
+				}
+				Transactor::on_deposited(what.clone(), who.clone(), Vec::new())
```
Yes, only xtransfer give those  callback

tolak commented on pallets/xtransfer/src/fungible_adapter.rs:
```diff
@@ -0,0 +1,118 @@
+					MultiLocation::new(0, X1(GeneralKey(b"bridge_transfer".to_vec())))
+						.into_account();
+				if NativeChecker::is_native_asset(what) {
+					NativeAdapter::deposit_asset(
```
It only determine whether we should forward, stuff done by `on_forwarded` callback function provided by xtransfer

tolak commented on pallets/xtransfer/src/chainbridge.rs:
```diff
@@ -0,0 +1,2408 @@
+	/// The common format is (31 bytes unique ID + 1 byte chain ID).
+	pub fn derive_resource_id(chain: u8, id: &[u8]) -> ResourceId {
+		let mut r_id: ResourceId = [0; 32];
+		r_id[31] = chain; // last byte is chain id
```
Unused function, removed

tolak commented on pallets/xtransfer/src/chainbridge.rs:
```diff
@@ -0,0 +1,2408 @@
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
```
Corrected

tolak commented on pallets/xtransfer/src/chainbridge.rs:
```diff
@@ -0,0 +1,2408 @@
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
```
Corrected

tolak commented:
> General questions:
> 1. How about changing `xtransfer` to `subbridge`?

Agree, renamed

> 2. How about calling the transfer bridge implementation something like "xcm_bridge" or "xcm_bridge_plugin" to imply it's one pluggable bridge in SubBridge?

Renamed to xcmbridge

> 3. What will happen if the bridge routing fails? Is it possible for the original sender chain to claim the stuck assets back?

So far we havn't handle this sitution when forwarding, i guess we need to impl something like `AssetTrap` and `AssetClaim` when doing forward. I have investigated polkadot XCM, they have `AssetTrap` and `AssetClaim` but not found how user use it. We can leave a TODO here, currently the un-forwarded assets will be saved in `temporary_account`, which is `MultiLocation::new(0, X1(GeneralKey(b"bridge_transfer".to_vec()))).into_account()`

> 4. Where is the received crosschain assets handled?

chainbridge::handle_fungible_transfer()

h4x3rotab reviewed (approved):
LGTM

h4x3rotab commented:
Let's rebase and merge?

tolak commented:
Will do it after migration test done
